daemons/ipa-kdb/ipa_kdb_principals.c

```diff
@@ -1587,6 +1607,12 @@ static krb5_error_code dbget_alias(krb5_context kcontext,
     if (krb5_realm_compare(kcontext, ipactx->local_tgs, norm_princ)) {
         /* In realm alias, try to retrieve it and let the caller handle it. */
         kerr = dbget_princ(kcontext, ipactx, norm_princ, flags, entry);
+    }
+
+    /* if we haven't found the principal in our realm, it might still
+     * be a referral to a known realm. Otherwise, bail out with the result */
+    if ((kerr != KRB5_KDB_NOENTRY) &&
+        (flags & KRB5_KDB_FLAG_REFERRAL_OK) == 0) {
```

Comment:
Based on the comment, my understanding is we should bail out if the principal entry is found, or if it is not found AND server referral is disabled. But this test is true only if the principal entry is not not found (i.e. found) and server referral is disabled.

Shouldn't it rather be something like:
```c
(kerr == 0) || ((kerr == KRB5_KDB_NOENTRY) && (flags & KRB5_KDB_FLAG_REFERRAL_OK) == 0))
```

Reply:
We might get other error conditions returned and we don't want to continue processing them. Hence, we only continue in the case we have not found an entry yet and the caller expects us to return a referral as well.

Reply:
But the test as it is means we will proceed regardless of the error in case server referral is enabled. Wouldn't it make more sense to return here in case the entry lookup succeed OR server referral is enabled?

Just after this block a comment says "The request is out of realm starting from here". The only case where we are sure this is the case is when the error code is `KRB5_KDB_NOENTRY`. However, we might reach this point with any error as long as server referral is enabled.

Reply:
The logic should be: "only continue if we asked to produce referrals and we haven't found anything. In case of any other error to find an entry or if we found an entry, return".

Reply:
So the condition to continue is:
```
(kerr == KRB5_KDB_NOENTRY) && ((flags & KRB5_KDB_FLAG_REFERRAL_OK) == 1)
```
Which means the condition for not continuing is:
```
(kerr != KRB5_KDB_NOENTRY) || ((flags & KRB5_KDB_FLAG_REFERRAL_OK) == 0)
```

Reply:
ok, I'll change to `||`.

Reply:
Actually, no. I changed to `||` and it crashed KDC due to returning NULL entry to the caller when attempting to get S4U2Self for AD user:
```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007f60a3b9c862 in ipadb_get_principal (kcontext=0x559c088287a0, search_for=<optimized out>, flags=80, entry=0x559c091367e8)
    at /usr/src/debug/freeipa-4.11.0.dev202305311005+gitd92b04935-0.fc39.x86_64/daemons/ipa-kdb/ipa_kdb_principals.c:1789
1789	    if (is_tgs_princ(kcontext, (*entry)->princ)) {
(gdb) print kerr
$1 = 0
(gdb) print *entry
$2 = (krb5_db_entry *) 0x0
```